---
title: "Analysis of the human mitochondrial proteome"
output: html_notebook
---

# Using Bioconductor to analyze mitochondrial proteome

## Installing a bunch of human packages from Bioconductor. 

(We can check which annotation packages are available on the [Bioconductor annotation package page](http://www.bioconductor.org/packages/release/data/annotation/)):

```{r, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite()
biocLite("EnsDb.Hsapiens.v86") #Load the latest version of human genome
biocLite("SNPlocs.Hsapiens.dbSNP144.GRCh38") #Load the lastest version of human SNPs
biocLite("BSgenome.Hsapiens.UCSC.hg38") #human genome
biocLite("TxDb.Hsapiens.UCSC.hg38.knownGene") #already installed
biocLite("KEGG.db") #not updated but used to get pathways IDs
biocLite("KEGGREST") #A package that provides a client interface to the KEGG REST server.
biocLite("org.Hs.eg.db") #an R object that contains mappings between Entrez Gene identifiers and GenBank accession numbers.
```

## Loading code libraries

```{r}
library(GenomicRanges)
library(rtracklayer)
library(BSgenome)
library(BiocInstaller)
```

## Loadin data libraries

Should we load them for individual sections of the code?

### Load the latest transcriptome for known genes (note, no mtDNA encoded genes):

```{r}
library(TxDb.Hsapiens.UCSC.hg38.knownGene) #notice, will load all the needed Bioconductor libraries
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
txdb
class(txdb)
```

### Retrieving genomic regions using `genes()`, `transcripts()`, `exons()`, `cds()`, and `promoters()`

```{r}
hs_all_gn_gr <- genes(txdb) #shows gene id as metadata
hs_all_cds_gr <- cds(txdb) #shows cds id as metadata
hs_all_tx_gr <- transcripts(txdb) #shows transcript id and transcript name
length(hs_all_gn_gr)
length(hs_all_cds_gr)
length(hs_all_tx_gr)
## Consult the documentation for this family of functions for extracting information from transcriptDb objects at `help(transcripts)`.
```

### Grouping GRangesList objects using `transcriptsBy()`, `exonsBy()`, `cdsBy()`, `intronsBy()`, 
`fiveUTRsByTranscript()`, and `threeUTRsByTranscript()`.

It’s often more natural to work with a GRangesList object of these types of features grouped by some other type of feature than working with a massive GRanges list object of everything. For example, we might want to retrieve all exons grouped by transcript or gene:

```{r}
hs_cds_by_gn <- cdsBy(txdb, by="gene")
length(hs_cds_by_gn)
head(hs_cds_by_gn, 100)
hs_cds_by_gn
```

See `help(transcriptsBy)` for more information.

## 1 Building gene lists

* lists for complete mitoproteome (mitocarta, mito)
* lists based on KEGG pathways
* list of mitochondrial (protein) genes

#### Proteome lists; hs_mitocarta_gn & hs_mtproteome_gn

**mtDNA genes**

```{r}
library(org.Hs.eg.db)
#columns(org.Hs.eg.db) #shows you what's there
help('select') #tells you how to access this db
allMTGenesOrgDb <- select(org.Hs.eg.db, keys=keys(org.Hs.eg.db), columns="CHR")
hs_all_gn <- allMTGenesOrgDb$ENTREZID  #all genes in annotation database
length(hs_all_gn)
hs_mtDNA_gn <- as.character(allMTGenesOrgDb$ENTREZID[allMTGenesOrgDb$CHR %in% "MT"])
```

**Mitocarta 2.0** [Broad institute website](https://www.broadinstitute.org/scientific-community/science/programs/metabolic-disease-program/publications/mitocarta/mitocarta-in-0) 1158 records

```{r}
hs_mitocarta_df <- read.csv(file = "Human.MitoCarta2.0.csv") 
hs_mitocarta_no_mtDNA_df <- hs_mitocarta_df[!hs_mitocarta_df$HumanGeneID %in% hs_mtDNA_gn,]
hs_mitocarta_gn <- hs_mitocarta_no_mtDNA_df$HumanGeneID[hs_mitocarta_no_mtDNA_df$HumanGeneID %in% hs_all_gn_gr$gene_id]
length(hs_mitocarta_gn)
## Mitocarta genes not in txdb
absent <- hs_mitocarta_no_mtDNA_df[!hs_mitocarta_no_mtDNA_df$HumanGeneID %in% hs_all_gn_gr$gene_id, ]
absent[c(2,4, 6)]
```

**Mitoproteome2** [http://www.mitoproteome.org/](http://www.mitoproteome.org/) 1705 records

```{r}
#hs_mtproteome_df <- read.delim(file = "mitoproteome.txt", header = TRUE, sep = "\t")
hs_mtproteome_df <- read.csv(file = "hs_mitoproteome.csv", header = TRUE)
nrow(hs_mtproteome_df)
tail(hs_mtproteome_df, 300)
absent <- hs_mtproteome_df[!hs_mtproteome_df$Entrez_Gene_ID %in% hs_mitocarta_df$HumanGeneID, ]
nrow(absent)
head(absent)
test1 <- hs_mtproteome_df[!hs_mtproteome_df$Entrez_Gene_ID %in% hs_mitocarta_gn$HumanGeneID, ] #135
```

**Mitominer4.0** [http://mitominer.mrc-mbu.cam.ac.uk/release-4.0/begin.do](http://mitominer.mrc-mbu.cam.ac.uk/release-4.0/begin.do)

```{r}
hs_mitominer_df  <- read.delim(file = "hs_mitominer4.0.tsv", header = TRUE, sep = "\t")
hs_mitominer_df ##1408
hs_mitocarta_gn
test2 <- hs_mitocarta_gn[!hs_mitocarta_gn$HumanGeneID %in% hs_mitominer_df$Gene.name, ] #179
test1[test1$HumanGeneID %in% test2$HumanGeneID, c(-1,-3, -5)]
```


#### KEGG lists

Such list can be created with the KEGG.db and KEGGgraph packages. Updates to the data in the KEGG.db 
package are no longer available, however, the resource is still useful for identifying pathway names 
and ids. Create a table of KEGG pathways and ids and search on the term ‘mitochondrial’.

```{r}
library(KEGG.db)
pathways <- toTable(KEGGPATHNAME2ID)
pathways[grepl("Oxidative phosphorylation", pathways$path_name, fixed=TRUE),]
head(pathways)
```

Use the ”00190” id to query the KEGG web resource (accesses the currently maintained data).
Similarly, can use "00020" to get the genes in Citrate cycle (TCA cycle)

> #############
> Tried KEGGgraph but got some strange results
> ##############

> ############ NOTE, WE NEED INTERNET FOR THESE; SHOULD STORE THESE LISTS SOMEWHERE #######

Defining two functions to use **KEGGREST** package

```{r}
gn_names_from_Kegg <- function(x, warn=TRUE) {
  #A function that returns a gene list based on KEGG pathway ID (e.g., hsa00190)
  library(KEGGREST) #A package that provides a client interface to the KEGG REST server.
  y <- keggGet(x)[[1]]$GENE[c(TRUE,FALSE)] # needed the T/F index to get read of gene description in alternative rows.
  return(y)
}

gn_df_from_Kegg <- function(x, warn=TRUE) {
  #A function that returns a dataframe with two columns: "gene_id" and "description" from the KEGG database based on pathway ID (e.g., hsa00190)
  library(KEGGREST) #A package that provides a client interface to the KEGG REST server.
  temp <- keggGet(x)[[1]]$GENE
  y <- data.frame(temp[c(TRUE,FALSE)],temp[c(FALSE,TRUE)])
  colnames(y) <- c("gene_id","description")
  return(y)
}
```

**Ox-phos genes** 

```{r}
hs_00190_gn <- gn_names_from_Kegg("hsa00190")
hs_00190_df <- gn_df_from_Kegg("hsa00190")
length(hs_00190_gn) # 133 genes
hs_00190_gn[!hs_00190_gn %in% hs_mitocarta_df$HumanGeneID] # 32 not in MitoCarta
hs_oxph_mc_gn <- hs_00190_gn[hs_00190_gn %in% hs_mitocarta_df$HumanGeneID] # 101 in MitoCarta
hs_00190_df[!hs_00190_df$gene_id %in% hs_mitocarta_df$HumanGeneID,] 
```

**aaRS** Notice, this set includes both mitochondrial and cytosolic aaRS as well as tRNAs
Transfer RNA biogenesis [BR:hsa03016]


```{r}
hs_00970_gn <- gn_names_from_Kegg("hsa00970")
hs_00970_df <- gn_df_from_Kegg("hsa00970")
length(hs_00970_gn)
hs_00970_gn
hs_00970_df
hs_00970_gn[!hs_00970_gn %in% hs_mitocarta_df$HumanGeneID] # 40 not in MitoCarta
hs_mt_aaRC_gn <- hs_00970_gn[hs_00970_gn %in% hs_mitocarta_df$HumanGeneID] # 26 in MitoCarta
hs_mt_aaRC_gn
hs_00970_df[hs_00970_df$gene_id %in% hs_mt_aaRC_gn,] # subsampling dataframe by genes in MitoCarta
```

Hmmmm, we can get the same set from Amphimedon!

```{r}
aq_00970_gn <- gn_names_from_Kegg("aqu00970")
aq_00970_df <- gn_df_from_Kegg("aqu00970")
length(aq_00970_gn)
aq_00970_gn
aq_00970_df
```


**TCA (Citrate cycle) genes**

```{r}
library(KEGGREST)
hs_00020_gn <- gn_names_from_Kegg("hsa00020") # needed the T/F index to get read of gene description in alternative rows.
hs_00020_df <- gn_df_from_Kegg("hsa00020")
length(hs_00020_gn) ## 30 genes
hs_00020_df
hs_00020_df[!hs_00020_df$gene_id %in% hs_mitocarta_df$HumanGeneID,] # 1 (5105) not in MitoCarta
```

**Pyruvate metabolism**

```{r}
hs_00620_gn <- gn_names_from_Kegg("hsa00620")
hs_00620_df <- gn_df_from_Kegg("hsa00620") #37
hs_00620_mc_df <- hs_00620_df[hs_00620_df$gene_id %in% hs_mitocarta_df$HumanGeneID,] #29
hs_00620_mc_df[!hs_00620_mc_df$gene_id %in% hs_00020_df$gene_id,] #10 in TCA cycle
```

**Mitochondrial biogenesis: protein import, information processing, and quality control (http://www.genome.jp/kegg-bin/get_htext?ko03029.keg) **

```{r}
hs_mt_import_df <- read.delim(file = "hs_protein_import.txt", header = TRUE, sep = "\t", blank.lines.skip = TRUE, comment.char = "#") 
hs_mt_import_df[!hs_mt_import_df$gene_id %in% hs_mitocarta_df$HumanGeneID, ] #12 not in mitocarta
hs_mt_dna_rna_df <-  read.delim(file = "hs_dna_rna.txt", header = TRUE, sep = "\t", blank.lines.skip = TRUE, comment.char = "#") 
hs_mt_dna_rna_df[!hs_mt_dna_rna_df$gene_id %in% hs_mitocarta_df$HumanGeneID, ] #13 not in mitocarta
hs_mt_quality_df <- read.delim(file = "hs_quality_control.txt", header = TRUE, sep = "\t", blank.lines.skip = TRUE, comment.char = "#") #183 in the list
hs_mt_quality_df[hs_mt_quality_df$gene_id %in% hs_mitocarta_df$HumanGeneID, ] #only 83 in mitocarta
```

**Apoptosis (hsa04210) genes in mitocarta**

```{r}
hs_04210_df <- gn_df_from_Kegg("hsa04210")
hs_04210_mc_df <- hs_04210_df[hs_04210_df$gene_id %in% hs_mitocarta_df$HumanGeneID, ]
hs_04210_mc_df
```

OK, saving these lists to an object so we don't depend on internet

```{r}
kegg_mt_df <- list(hs_00020_df, hs_00190_df, hs_04210_df)
names(kegg_mt_df) <- c("TCA", "OxPhos", "Apoptosis")
save(kegg_mt_ls, file="kegg_df.rda")
load(file = "kegg_df.rda") #if we want to reload them
```

**Mitochondrial ribosomal proteins (47)**
Was not quite sure how to load them from KEGG, used lists at http://www.genome.jp/kegg-bin/get_htext#C4

```{r}
hs_ssrp_gn <- read.csv(file="hs_ssrp.txt", header = FALSE)
colnames(hs_ssrp_gn) <- c("GeneID","ProtID","Description")
hs_lsrp_gn <- read.csv(file="hs_lsrp.txt", header = FALSE)
colnames(hs_lsrp_gn) <- c("GeneID","ProtID","Description")
head(hs_lsrp_gn)
hs_mitocarta_df[hs_mitocarta_df$HumanGeneID %in% hs_lsrp_gn$GeneID, ]
```

**Mitochondrial carriers**
A superfamily of transporters named SLC25 in mammals, having a particular tripartite structure.
Humans have 53 of them: Palmieri F. (2013). 

```{r}
hs_mito_carrier_id <- read.csv(file="hs_mito_carriers.cds", header = FALSE)
hs_mito_carrier_id <- as.vector(t(hs_mito_carrier_id))
hs_mito_carrier_gn <- sort(unique(names_parsed_df[names_parsed_df$gene_name %in% hs_mito_carrier_id, 3]))
hs_mito_carrier_gn
```




## 2 Collecting CDS and Protein sequences for a set of genes

### 2A Genome annotation approach

Note that sections 2 and 3 below explain individual steps in this procedure

```{r}
get_hs_cds <- function(x, warn=TRUE) {
  library(TxDb.Hsapiens.UCSC.hg38.knownGene)
  library(BSgenome.Hsapiens.UCSC.hg38)
  txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
  hs_gm <- BSgenome.Hsapiens.UCSC.hg38
  x <- as.character(x[x %in% hs_all_gn_gr$gene_id]) #only use genes for which we have info in txdb
  
  #### Granges for CDSs and UTRs
  txbygene <- transcriptsBy(txdb, "gene")[x] ## genetic ranges for genes of interest
  map <- relist(unlist(txbygene, use.names=FALSE)$tx_id, txbygene) 
  # Converts a GrangesList object with ranges, chr, strands, tx_name into a list of tx_ids for each gene
  # relist(flesh, skeleton), where flesh is a vector-like object and skeleton is a list-like object. 
  # Only the "shape" (i.e. the lengths of the individual list elements) of skeleton matters. Its exact content is ignored.
  
  #### Extract the UTR and coding regions for all(!) genes.
  cds <- cdsBy(txdb, "tx")
  #threeUTR <- threeUTRsByTranscript(txdb) #don't use it for now
  #fiveUTR <- fiveUTRsByTranscript(txdb)   #don't use it for now
  
  #### Get the UTR and coding regions for selected genes (listed in map).
  txid <- unlist(map, use.names=FALSE)
  cds <- cds[names(cds) %in% txid]
  
  #### Using `extractTranscriptSeqs` to extract the UTR and coding regions from the BSgenome. 
  cds_seqs <- extractTranscriptSeqs(hs_gm, cds)
  #threeUTR_seqs <- extractTranscriptSeqs(hs_gm, threeUTR) #don't use it for now
  #fiveUTR_seqs <- extractTranscriptSeqs(hs_gm, fiveUTR)   #don't use if for now
  
  #### Collecting the coding and UTR regions into groups by gene id. 
  lstc <- relist(cds_seqs, PartitioningByWidth(sum(map %in% names(cds)))) # sum(map %in% names(cds)) how many seq in each category
  #lst3 <- relist(threeUTR_seqs, PartitioningByWidth(sum(map %in% names(threeUTR)))) #don't use it for now
  #lst5 <- relist(fiveUTR_seqs, PartitioningByWidth(sum(map %in% names(fiveUTR))))   #don't use it for now
  return(lstc)
}
```

Try it:

```{r}
hs_dna_rna_seq <- get_hs_cds(hs_mt_dna_rna_df$gene_id)
prots <- lapply(hs_dna_rna_seq, translate)
tryme <- unlist(prots, use.names=TRUE)
tryme
writeXStringSet(tryme, file="hs_mt_dna_rna_cds.seq", format="fasta")
```

### 2B BIOMARTR approach

Notice that there is a more mature BIOMART package, but it is not the one we are using here (although it is one of the dependencies)

#### Loading first

```{r}
library(biomartr)
listMarts()
ensembl <- useMart("ensembl")
listDatasets(ensembl)
ensembl <-  useMart("ensembl", dataset="hsapiens_gene_ensembl")
snpmart  <-  useMart(biomart = "ENSEMBL_MART_SNP", dataset="hsapiens_snp")
```

#### Downloading CDS
The getCDS() function is an interface function to the NCBI RefSeq, NCBI Genbank, ENSEMBL and ENSEMBLGENOMES databases from which corresponding CDS files can be retrieved. It works analogous to getGenome() and getProteome().

Options are:

* db = "refseq" for retrieval from NCBI RefSeq
* db = "genbank" for retrieval from NCBI Genbank
* db = "ensembl" for retrieval from ENSEMBL
* db = "ensemblgenomes" for retrieval from ENSEMBLGENOMES

Furthermore, again users need to specify the scientific name of the organism of interest for which a proteomes shall be downloaded, e.g. organism = "Homo sapiens". Finally, the path argument specifies the folder path in which the corresponding CDS file shall be locally stored. 

#### Downloading RefSeq CDS:
- download the genome of Homo sapiens from refseq
- and store the corresponding genome CDS file in '_ncbi_downloads/CDS'

```{r}
HS.cds.refseq <- getCDS( db       = "refseq", 
                         organism = "Homo sapiens",
                         path     = file.path("data","_ncbi_downloads","CDS"))
human_cds_bs <- read_cds(file     = HS.cds.refseq, 
                      obj.type = "Biostrings")

### Parsing the names

hs_cds_names <- names(human_cds_bs)
names_parsed <- sub(".*(NC_[^ ]+) .*gene=(\\w+).*GeneID:(\\w+).*protein=([^\\]]+).*protein_id=([^\\]]+).*location=([^\\]]+).*", "\\1,,\\2,,\\3,,\\4,,\\5,,\\6", hs_cds_names, perl=TRUE) ## had to add to ,, to avoid splitting location
names_parsed_df <- as.data.frame(do.call(rbind, strsplit(names_parsed, ',,')),stringsAsFactors=FALSE)
names(names_parsed_df) <- c("cds_id","gene_name","gene_id","protein_name","protein_id","location")
```


#### Downloading RefSeq Proteome


```{r}
hs_pep_refseq <- getProteome( db       = "refseq", 
                         organism = "Homo sapiens",
                         path     = file.path("data","_ncbi_downloads","Proteome"))
hs_proteins <- read_proteome(file = hs_pep_refseq, obj.type = "Biostrings")

```

#### Extracting specific set of proteins

```{r}
### Defining a function to get ref protein sequences for a set of genes
get_hs_ref_pep <- function(x, warn=TRUE) {
  prot_names <- names(hs_proteins)
  prot_id <- sub(" .*", "", prot_names) #delete everything after the first space
  ## need parsed names of cds sequences (they have gene_id and protein_id info)
  to_extract <- sort(names_parsed_df$protein_id[names_parsed_df$gene_id %in% x])
  prot_set <- hs_proteins[prot_id %in% to_extract, ]
  ## Removing duplicates
  prot_set <- prot_set[!duplicated(prot_set)]
  return(prot_set)
}
```

Checking it:

```{r}
hs_mit_dna_rna_pep <- get_hs_ref_pep(hs_mt_dna_rna_df$gene_id)
writeXStringSet(hs_mit_dna_rna_pep, file="hs_mt_dna_rna_pep.seq", format="fasta")
hs_00020_pep <- get_hs_ref_pep(hs_00020_gn)
writeXStringSet(hs_00020_pep, file="hs_00020_pep.seq", format="fasta")
hs_mito_carrier_pep <- get_hs_ref_pep(hs_mito_carrier_gn)
hs_mito_carrier_pep
writeXStringSet(hs_mito_carrier_pep, file="hs_mito_carrier_pep.seq", format="fasta")
```



### Handling Many Sequences with XStringSet

The XStringSet class allows to represent many sequences in a single container. The subclasses for the different types of biosequences are: BStringSet, RNAStringSet, DNAStringSet and AAStringSet. 

```{r}
dset <- DNAStringSet(Human_CDS) ## This actually worked!  Looks much better now.
dset_names <- names(dset)
```

Pooling out every sequence that has "mitochondrial" in its name:

```{r}
mset <- dset[grep("mitochondrial", names(dset))] ## Returned a DNAStringSet instance of length 2906 (but only 622 genes)
head(names(mset))
```

Pooling out sequences corresponding to geneIDs in Mitocarta. Checking for obsolete sequences in mitocarta **Found 13**

```{r}
dset_names <- names(dset)
dset_names
dset_gID <- sub(".*GeneID:", "", dset_names)
dset_gID <- sub("].*", "", dset_gID)
head(dset_gID)
## pooling out all dset CDS sequences that have corresponding GeneID in mitocarta
mt_ref_set <- dset[dset_gID %in% Human.MitoCarta2.0$HumanGeneID]
## pooling out all MitoCarta2.0 sequences without corresponding Human refseq
check_mitocarta <- Human.MitoCarta2.0[!Human.MitoCarta2.0$HumanGeneID %in% dset_gID,]
sort(check_mitocarta$HumanGeneID)
# 13 mitocarta genes are not in Human refseq anymore: 4508 4509 4512 4513 4514 4519 4535 4536 4537 4538 4539 4540 4541
```

Getting proteins:



```{r}
my_ids <- hs_mt_dna_rna_df$gene_id
sort(my_ids) #74 ids
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
protein = getSequence(id=my_ids,
                      type="entrezgene",
                      seqType="peptide", 
                      mart=ensembl)

```



### 3. Identify genomic coordinates

The list of gene ids is used to extract genomic positions of the regions of interest. 

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene  #done before
hs_all_gn_gr <- genes(txdb) #shows gene id as metadata #25017 genes
```


#### Obtaining GRanges for some of our lists

First defining a simple function

```{r}
hs_granges4gn <- function(x) {
  ## Requires #library(TxDb.Hsapiens.UCSC.hg38.knownGene)
  txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene  #done before
  hs_all_gn_gr <- genes(txdb) #shows gene id as metadata 
  y <- hs_all_gn_gr[hs_all_gn_gr$gene_id %in% x]
  return(y)
}
```

Obtaining GRanges

```{r}
### MITOCARTA ###
hs_mitocarta_gr <- hs_granges4gn(hs_mitocarta_df$HumanGeneID)
length(hs_mitocarta_gr)
### 1109 hmmm

### TXDB genes not in MITOCARTA ###
hs_nuc_gr <- hs_all_gn_gr[!hs_all_gn_gr$gene_id %in% hs_mitocarta_df$HumanGeneID]
length(hs_nuc_gr)

### OX-PHOS ###
hs_op_gr <- hs_granges4gn(hs_00190_gn)
length(hs_op_gr)

### TCA ###
hs_tca_gr <- hs_granges4gn(hs_00020_gn)
length(hs_tca_gr)
```

### 4. Finding chromosomal affiliation

Using GenomicFeatures for extracting subsets of features that overlap a specific 
chromosome or range.

```{r}
#seqlevels(txdb)
standard_chromosomes <- paste("chr",1:22,sep = "")
standard_chromosomes <- append(standard_chromosomes, c("chrX", "chrY"))
seqlevels(txdb, force=TRUE) <- standard_chromosomes
seqlevels(txdb)
#txdb <- restoreSeqlevels(txdb) # restore txdb so it queries all sequences
```

```{r}
# Calculating and plotting a number of genes per chromosome (and chromosome length)
## optional, if you want a genomic order of the chromosomes
genesPerChromosome <- function (x, warn=TRUE) {
  ## A function that maps genes to chromosomes
  x = sortSeqlevels(x)
  seqlevels(x, force=TRUE) <- standard_chromosomes
  ## split into a GRangesList
  ## where each element has all ranges for one chromosome
  grl = split(x, seqnames(x))
  ## apply a function to the ranges of each chromosome
  print(sapply(grl, length))
  sum(sapply(grl, length))
  len = round(seqlengths(grl)/1000000)
  df <- data.frame(c(sapply(grl, length)), len)
  df <- df[c(2,1)]
  #### something weird: it does not work the other way around:
  #df <- data.frame(c(seqlengths(grl):1000000, sapply(grl, length)))
  colnames(df)=c("chr_length (Mb)","gene_num")
  plot(df)
}
genesPerChromosome(hs_op_gr)
genesPerChromosome(hs_mitocarta_gr)
```

#######
Should we be surprised that all chromosomes have genes involved in OX_Phos?
#######


### 5. Plotting chromosomal locations using chromPlot package

```{r}
biocLite("chromPlot")
library("chromPlot")
data(hg_gap)
head(hg_gap)
chromPlot(gaps=hg_gap, annot1=hs_op_gr)
chromPlot(gaps=hg_gap, annot1=hs_tca_gr)
chromPlot(gaps=hg_gap, bands=hg_cytoBandIdeo, annot1=hs_mitocarta_gr, figCols=6)
```



#### Retrieving genomic regions for mitochondiral genes from human genome

```{r, eval=FALSE}
library(BSgenome)
library(BSgenome.Hsapiens.UCSC.hg38)
```

This is a BSgenome package, where BS stands for Biostrings, a Bioconductor package that contains classes 
for storing sequence data and methods for working with it. BSgenome packages contain the full reference 
genome for a particular organism, compressed and wrapped in a user-friendly package with common accessor 
methods. As always, it’s worth reading the vignettes for these packages on Bioconductor’s website.

```{r}
hs_gm <- BSgenome.Hsapiens.UCSC.hg38
organism(hs_gm)
providerVersion(hs_gm)
provider(hs_gm)
```

```{r}
str(hs_gm)
seqinfo(hs_gm)
```

### Retrieving the regions

```{r}
mcarta_seq <- getSeq(hs_gm, cds)
mcarta_seq
```


Wow, that was easy!

```{r}
write.csv(hs_mtproteome_df, file="hs_mitoproteome.csv")

```

## Trying BLAST

```{r}
library("devtools", lib.loc="/Users/dlavrov/Library/R/3.3/library")
install_github("mhahsler/rBLAST")
library(rBLAST)

library(annotate)
# search using a fragment of TLR4 mRNA.
x <- blastSequences("acttcttggg cttagaacaa ctagaacatc tggatttcca gcattccaat ttgaaacaaa")
x

### Here is the alternative:
myPipe <- pipe( "blastall -p blastp -i text.fasta -d data.fasta" )
results <- read.table( myPipe )
colnames( blastResults ) <- c( "QueryID",  "SubjectID", "Perc.Ident",
"Alignment.Length", "Mismatches", "Gap.Openings", "Q.start", "Q.end",
"S.start", "S.end", "E", "Bits" )

### ONE MORE
seqs <-
  ## The sequences to be blasted and their fasta 'deflines' as an character vector.  
  c('>s1','CTTTGTCTTCCCCTATGTCATCCTCAATCTCTATGAAAGCAACAC'
    ,'>s2','CTATGTCATCCTCAATCTCTATGAAAGCAACACCGCTACCATAGA')

blast.f6 <-
  ## The fields you want back from blast.  c.f. `blastn -help` for a full list of available fields.
  c('qseqid', 'sseqid', 'pident', 'qcovs')

blast.out <-
  ## Run the job, converting output to an R data.frame
  system2('blastn',c('-db',"'nt'"
                    ,'-outfmt',sprintf('"6 %s"',paste(collapse=' ',blast.f6))
                    ,'-perc_identity',"'.90'"
                    ,'-entrez_query','"Aichi virus 1[ORGANISM]"'
                    ,'-remote'
                     )
         ,input=seqs
         ,stdout=TRUE                   # capture the output in a character vector
          )

blast.out.df <-
  ## parse blast.out as a table and assign names to it
  `names<-`(read.table(quote=""
                      ,sep='\t'
                      ,textConnection(blast.out)
                       )
           ,blast.f6)


```



